<?xml version="1.0" encoding="utf-8" standalone="yes" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
  <channel>
    <title>エラー解決 on プログラマー備忘録</title>
    <link>https://sankame.github.io/categories/%E3%82%A8%E3%83%A9%E3%83%BC%E8%A7%A3%E6%B1%BA/index.xml</link>
    <description>Recent content in エラー解決 on プログラマー備忘録</description>
    <generator>Hugo -- gohugo.io</generator>
    <language>ja</language>
    <copyright>Powered by [Hugo](//gohugo.io). Theme by [hugo-theme-wave](https://github.com/syui/hugo-theme-wave).</copyright>
    <atom:link href="https://sankame.github.io/categories/%E3%82%A8%E3%83%A9%E3%83%BC%E8%A7%A3%E6%B1%BA/index.xml" rel="self" type="application/rss+xml" />
    
    <item>
      <title>VirtualBox起動エラー(rc=-22838)</title>
      <link>https://sankame.github.io/blog/2019-10-05-virtualbox_does_not_start/</link>
      <pubDate>Sat, 05 Oct 2019 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2019-10-05-virtualbox_does_not_start/</guid>
      <description>

&lt;p&gt;4ヶ月くらい起動していなかったVirtualBoxで久々に開発をしようと思ったらエラーが発生しました。
ネット上の情報はあまりなく、下記のいずれかでした。Bはいつの時点に戻すか分からなかったので、Aでやることにしました。&lt;/p&gt;

&lt;p&gt;(A)VirtualBoxを再インストール。&lt;/p&gt;

&lt;p&gt;(B)Windowsの復元機能を使って、Windowsアップデート前の状態にする。&lt;/p&gt;

&lt;h1 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h1&gt;

&lt;p&gt;&lt;code&gt;[OS] Windows7 (32bit)&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[VirtualBox] 4.3.26 -&amp;gt; 5.2.32&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Vagrant] 1.7.2 -&amp;gt; 2.2.5&lt;/code&gt;&lt;/p&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;VagrantからVirtualBoxを起動する構成です。(&lt;a href=&#34;https://sankame.github.io/blog/2015-05-06-vagrant_setup/&#34; target=&#34;_blank&#34;&gt;詳細はこちら&lt;/a&gt;)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;いつものように&lt;code&gt;vagrant up&lt;/code&gt;や&lt;code&gt;vagrant resume&lt;/code&gt;でVirtualBoxを起動しようとしたところ、このエラーが出ました。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/virtualbox_does_not_start/error_22838.png&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;60%&#34; height=&#34;60%&#34; title=&#34;error_22838&#34; alt=&#34;error_22838&#34; src=&#34;https://sankame.github.io/assets/posts/img/virtualbox_does_not_start/error_22838.png&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;エラーメッセージ&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-text&#34;&gt;supHardenedWinVerifyProcess failed with VERR_ASN1_CURSOR_TAG_MISMATCH: (rc=-22838)
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-対応&#34;&gt;2. 対応&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;内部エラーで複雑そうなので、一度VirtualBoxをアンインストールし、新しいバージョンをインストールしました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;32bit OSに対応してて、新し目のバージョン5.2.32を選択しました。(&lt;a href=&#34;https://www.virtualbox.org/wiki/Download_Old_Builds_5_2&#34; target=&#34;_blank&#34;&gt;5.2.32のWindows hostsをクリック&lt;/a&gt;)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;64bit OSを使ってる方は、もっと新しいバージョンの方がいいと思います。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;VirtualBoxを再インストール後、&lt;code&gt;vagrant up&lt;/code&gt;を実行すると下記エラーメッセージが表示されました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;エラーメッセージ&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-text&#34;&gt;The provider &#39;virtualbox&#39; that was requested to back the machine
&#39;default&#39; is reporting that it isn&#39;t usable on this system. The
reason is shown below:

Vagrant has detected that you have a version of VirtualBox installed
that is not supported. Please install one of the supported versions
listed below to use Vagrant:

4.0, 4.1, 4.2, 4.3
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;現在のVagrantのバージョン(1.7.2)だと、上記バージョンのVirtualBoxにしか対応できないようです。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;VirtualBoxをバージョンダウンするより、Vagrantを上げた方がいいだろうと思って、一度Vagrantをアンインストールした後、下記から最新版を落としてきてインストールしました。(該当するOSとアーキテクチャのリンクからダウンロード。&lt;a href=&#34;https://www.vagrantup.com/downloads.html&#34; target=&#34;_blank&#34;&gt;Download Vagrant&lt;/a&gt;)&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;Vagrantを再インストール後、再び&lt;code&gt;vagrant up&lt;/code&gt;するとまたエラーが出ました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;エラーメッセージ&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-text&#34;&gt;Vagrant failed to initialize at a very early stage:
        
The version of powershell currently installed on this host is less than the required minimum version. Please upgrade the installed version of powershell to the minimum required version and run the command again.

Installed version: 2

Minimum required version: 3
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;Vagrantをバージョンアップしたら今度はpowershellが対応できなくなったようです・・・&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;powershell ver3以上ということなので、下記の記事を参考にver4をインストールしました。(参考記事3-2、 3-3)&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;参考記事3-3のインストール・ウィザードに関して、僕の環境は32bit OSだったので、Windows6.1-KB2819745-x86-MultiPkg.msuを選択。(64bit版の場合、Windows6.1-KB2819745-x64-MultiPkg.msu)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;これでようやく復旧しました。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;3-参考にした記事&#34;&gt;3. 参考にした記事&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;3-1. &lt;a href=&#34;https://forums.virtualbox.org/viewtopic.php?f=6&amp;t=75160&#34; target=&#34;_blank&#34;&gt;Failed to start session for virtual machine&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;3-２. &lt;a href=&#34;https://www.whyit.work/entry/2018/04/20/154907&#34; target=&#34;_blank&#34;&gt;Windows PowerShellをバージョンアップする&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;3-３. &lt;a href=&#34;https://qiita.com/takeru08ma/items/969d6656be0086e9242b&#34; target=&#34;_blank&#34;&gt;Windows Power Shell 4.0 を Windows7 にインストールする手順&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>【Androidアプリ開発】CursorWindowAllocationException エラー</title>
      <link>https://sankame.github.io/blog/2018-08-05-android_cursor_error/</link>
      <pubDate>Sun, 05 Aug 2018 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2018-08-05-android_cursor_error/</guid>
      <description>

&lt;p&gt;CursorWindowAllocationException が発生した時の対処方法です。&lt;/p&gt;

&lt;h1 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h1&gt;

&lt;p&gt;&lt;code&gt;[Windows] 7&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Java] 1.8&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Android] 5.1.1&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Android API] 26&lt;/code&gt;&lt;/p&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;開発中のAndroidアプリでSQLiteを使ってます。&lt;/p&gt;

&lt;p&gt;サブクエリを使って、今までより抽出データが多くなるSQLを書いたところ、画面遷移中に頻繁にクラッシュするようになりました。&lt;/p&gt;

&lt;p&gt;しかも、特定のテスト機(Galaxy S4(Android5.0.1))のみです。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;(A)エラー内容&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;Fatal Exception: android.database.CursorWindowAllocationException
Cursor window allocation of 2048 kb failed. # Open Cursors=700 (# cursors opened by this proc=700)
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;(B)エラー箇所&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;前述のサブクエリを適用した箇所とは全く別の箇所が毎回落ちるポイントでした。(下記2行目)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-java&#34;&gt;Cursor cursorCountUsers = db.rawQuery(&amp;quot;SELECT COUNT(user_id) FROM users&amp;quot;, null);
cursorCountUsers.moveToFirst();
int countUserId = cursorCountUsers.getInt(0);
return countUserId;
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;&lt;a href=&#34;https://stackoverflow.com/questions/21219039/android-database-cursorwindowallocationexception-when-moving-a-cursor&#34; target=&#34;_blank&#34;&gt;Slackoverflow&lt;/a&gt;で同じ問題を見つけました。&lt;/p&gt;

&lt;p&gt;まず、毎回同じ箇所で落ちるので、上記(B)に&lt;code&gt;cursorCountUsers.close();&lt;/code&gt;を入れたけど直りませんでした。&lt;/p&gt;

&lt;p&gt;次に、ソースコード全体を&amp;rdquo;Cursor&amp;rdquo;で検索をかけ、closeしてない箇所が20くらいあったので全てに追記したところ問題は解消されました。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;Javaはgcがあると思ってcloseを甘く見てましたね・・・気を付けないと。&lt;/p&gt;

&lt;p&gt;RAMの空きが30数%あるのに落ちるということは、Slackoverflowの回答者が言う通り、cursorのために用意されたメモリ容量に上限があって、今までよりメモリを使うSQLを書いたせいで、そこが圧迫されたということかもしれません。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>【Androidアプリ開発】FileUriExposedException エラー</title>
      <link>https://sankame.github.io/blog/2018-07-23-android_use_fileprovider/</link>
      <pubDate>Mon, 23 Jul 2018 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2018-07-23-android_use_fileprovider/</guid>
      <description>

&lt;p&gt;Android7 で android.os.FileUriExposedException が発生した時の対処方法です。&lt;/p&gt;

&lt;p&gt;(文中の【】は読み替えて下さい)&lt;/p&gt;

&lt;h1 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h1&gt;

&lt;p&gt;&lt;code&gt;[Windows] 7&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Java] 1.8&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Android] 5.1.1&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;[Android API] 26&lt;/code&gt;&lt;/p&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Androidアプリから一時ファイルを作成して、そのURIを外部のアプリに渡すという処理を実装してました。&lt;/p&gt;

&lt;p&gt;(ファイルを作成して、Gmailに添付するなど)&lt;/p&gt;

&lt;p&gt;Android7からfile://による指定ができなくなったため、下記のエラーが発生します。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-java&#34;&gt;android.os.FileUriExposedException: file:///storage/emulated/0/Android/data/【アプリのパッケージ名】/temporary/【一時ファイル名】 exposed beyond app through ClipData.Item.getUri()
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;URIはFileProviderで作らないといけないそうです。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h1&gt;

&lt;h2 id=&#34;2-1-androidmanifest-xml&#34;&gt;2-1. AndroidManifest.xml&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;まず、AndroidManifest.xmlにFileProviderの定義を追加します。&lt;/p&gt;

&lt;p&gt;&lt;code&gt;android:resource&lt;/code&gt;には、2-2で作成するファイル名を指定。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;script src=&#34;https://gist.github.com/Sankame/9a8e44313453ca63dcdc18870bfd1179.js&#34;&gt;&lt;/script&gt;

&lt;h2 id=&#34;2-2-res-xml-provider-paths-xml&#34;&gt;2-2. res/xml/provider_paths.xml&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;次に、res/xml/provider_paths.xml を作成します。&lt;/p&gt;

&lt;p&gt;後述の2-3でfileの保存先を &lt;code&gt;Environment.getExternalStorageDirectory()&lt;/code&gt; の直下にしたので、それを表す&lt;code&gt;external-path&lt;/code&gt;を指定します。&lt;/p&gt;

&lt;p&gt;また、直下なので&lt;code&gt;path&lt;/code&gt;はカレントディレクトリになります。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;script src=&#34;https://gist.github.com/Sankame/6cee8d95f46725bf967ef207e13d2ba8.js&#34;&gt;&lt;/script&gt;

&lt;h2 id=&#34;2-3-該当処理を行っているactivity&#34;&gt;2-3. 該当処理を行っているActivity&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;最後に、URIの生成を下記(1)から(2)に変更しました。&lt;/li&gt;
&lt;/ul&gt;

&lt;script src=&#34;https://gist.github.com/Sankame/67c40ef8056fba82c23937e2ab54b161.js&#34;&gt;&lt;/script&gt;

&lt;h1 id=&#34;3-参考&#34;&gt;3. 参考&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;https://developer.android.com/reference/android/support/v4/content/FileProvider&#34; target=&#34;_blank&#34;&gt;FileProviderリファレンス&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>【Androidアプリ開発】 &#34;Socket is closed&#34; エラー</title>
      <link>https://sankame.github.io/blog/2018-05-07-android_socket_is_closed/</link>
      <pubDate>Mon, 07 May 2018 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2018-05-07-android_socket_is_closed/</guid>
      <description>

&lt;h1 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h1&gt;

&lt;pre&gt;&lt;code&gt;[Windows] 7
[Java] 1.8
[Android] 5.1.1
[Android API] 26
&lt;/code&gt;&lt;/pre&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Androidアプリの開発で、自前のAPIからデータを取得する処理を組んでいたところ、下記エラーが発生するようになりました。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-java&#34;&gt;java.net.SocketException: Socket is closed
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;コードはこちら。&lt;/p&gt;

&lt;script src=&#34;https://gist.github.com/Sankame/848d54b5e12f210811db5b80f8512553.js&#34;&gt;&lt;/script&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;まずAPI側でApacheのログを確認したところ、ステータスコード200で問題なしでした。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-apache&#34;&gt;127.0.0.1 - - [05/May/2018:22:55:02 +0900] &amp;quot;POST /api/test HTTP/1.0&amp;quot; 200 11130
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;正常終了するケースとの違いは、レスポンスのデータ量が多くなってることかな。(上記ログ右端)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;次にAndroidアプリ側の処理があやしいと考えて、いろいろ試した結果、HttpsURLConnectionのクローズのタイミングが問題だと分かりました。&lt;/p&gt;

&lt;p&gt;同オブジェクトを参照しているオブジェクトを全て使い終わってから、クローズする必要がありそうです。(上記(1)→(2)に移す)&lt;/p&gt;

&lt;p&gt;また、クローズ処理を消すだけでも冒頭のエラーは出なくなりましたが、なるべくオブジェクトは明示的に閉じた方が良いと思います。&lt;/p&gt;

&lt;p&gt;なので、最終的にオブジェクトの作成順とは逆にクローズしていく処理を入れました。(下記(3)(4))&lt;/p&gt;

&lt;script src=&#34;https://gist.github.com/Sankame/561d4244e309efd1b9f8f4c2aa72e714.js&#34;&gt;&lt;/script&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>ERROR 2002 (HY000): Can&#39;t connect to local MySQL server through socket &#39;/tmp/mysql.sock&#39; (38)</title>
      <link>https://sankame.github.io/blog/2017-11-20-fail_to_connect_mysql/</link>
      <pubDate>Mon, 20 Nov 2017 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2017-11-20-fail_to_connect_mysql/</guid>
      <description>

&lt;h1 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h1&gt;

&lt;pre&gt;&lt;code&gt;[Mac] 10.11.6
[MySQL] 5.7.20
&lt;/code&gt;&lt;/pre&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;MacにMySQLをインストール後、接続しようとするとエラーが出ました。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;$ brew install mysql

$ mysql -uroot
ERROR 2002 (HY000): Can&#39;t connect to local MySQL server through socket &#39;/tmp/mysql.sock&#39; (38)
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;ネットの情報をヒントに、エラーが出ているファイルの所有者を変えたら直りました。(ログインユーザー名は読み替えて下さい)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;$ ll /tmp/mysql.sock
-rw-r--r--  1 root  wheel  0 11 19 23:46 /tmp/mysql.sock

$ sudo chown [ログインユーザー名] /tmp/mysql.sock
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;念のためリスタート。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;$ brew services restart mysql
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;接続できました。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;$ mysql -uroot
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>【Androidアプリ開発】onPrepareActionModeが呼ばれない</title>
      <link>https://sankame.github.io/blog/2017-10-17-android_onprepareactionmode_not_called/</link>
      <pubDate>Tue, 17 Oct 2017 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2017-10-17-android_onprepareactionmode_not_called/</guid>
      <description>

&lt;h2 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h2&gt;

&lt;pre&gt;&lt;code&gt;[Windows] 7
[Java] 1.8
[Android] 5.1.1
[Android API] 26
&lt;/code&gt;&lt;/pre&gt;

&lt;h2 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;build.gradleで下記の値を上げたところ、onPrepareActionModeが実行されなくなってました。&lt;/p&gt;

&lt;p&gt;onPrepareActionModeは、ListViewのロングタップ時にアクションモードへ切り替えた後、アクションバー内のあるアイコンを隠す処理で使っていたので、困りました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;compileSdkVersion : 21 → 26&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;buildToolsVersion : 21.1.2 → 26.0.2&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h2 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;試しに下記 (A) にあった処理を (B) へ移してみたところ、思った通りの動きをしてくれました。&lt;/p&gt;

&lt;p&gt;うーん、前はダメだったからonPrepareActionModeを使ったんだけどなあ。。&lt;/p&gt;

&lt;p&gt;他にネットで情報を見つけられなかったので、一旦これにしておくか。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-java&#34;&gt;list1 = findViewById(R.id.listView1);

list1.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener() {

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        //(A)
        //MenuItem menuItem1 = menu.findItem(R.id.menu_item1);
        //menuItem1.setVisible(false);
        return false;
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_action_mode, menu);

        //(B)
        MenuItem menuItem1 = menu.findItem(R.id.menu_item1);
        menuItem1.setVisible(false);

        return true;
    }
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>ApacheのProxyPassでhtmlタグがそのまま表示される</title>
      <link>https://sankame.github.io/blog/2016-12-25-mod_proxy_text_plain/</link>
      <pubDate>Sun, 25 Dec 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-12-25-mod_proxy_text_plain/</guid>
      <description>

&lt;h3 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h3&gt;

&lt;pre&gt;&lt;code&gt;[CentOS] 6.8 (64bit版カーネル)
[Apache] 2.2.15
&lt;/code&gt;&lt;/pre&gt;

&lt;h3 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;ApacheのProxyPassディレクティブ(mod_proxy)を指定して、バックエンドから受け取ったコンテンツを表示しようとするとhtmlタグがそのまま表示されてしまいます。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Apache設定ファイルの抜粋&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-apache&#34;&gt;NameVirtualHost *:80
        
# 省略
        
&amp;lt;VirtualHost *:80&amp;gt;
    ServerAdmin test@local
    ServerName test.local

    ProxyPreserveHost On
    ProxyPass / http://127.0.0.1:8080/
    ProxyPassReverse / http://127.0.0.1:8080/

    ErrorLog logs/test.local-error_log
    CustomLog logs/test.local-access_log combined
&amp;lt;/VirtualHost&amp;gt;
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;ChromeのDeveloper Toolsなどで見ると、下記のようにHTTPレスポンスヘッダーで平文が指定されています。&lt;/p&gt;

&lt;p&gt;&lt;code&gt;Content-Type text/plain; charset=UTF-8&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;デフォルトのMIMEタイプ指定を無効にすることで解消しました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Apache設定ファイルの抜粋&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-apache&#34;&gt;NameVirtualHost *:80
        
# 省略
        
&amp;lt;VirtualHost *:80&amp;gt;
    ServerAdmin test@local
    ServerName test.local

    # ----- この行を追加 ----- 
    DefaultType None
    # ----- この行を追加 ----- 

    ProxyPreserveHost On
    ProxyPass / http://127.0.0.1:8080/
    ProxyPassReverse / http://127.0.0.1:8080/

    ErrorLog logs/test.local-error_log
    CustomLog logs/test.local-access_log combined
&amp;lt;/VirtualHost&amp;gt;
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;3-参考ページ&#34;&gt;3. 参考ページ&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;http://stackoverflow.com/questions/5034131/mod-proxy-ajp-error-renders-html-as-text-plain-prompts-user-to-save-as&#34; target=&#34;_blank&#34;&gt;mod_proxy_ajp error: renders html as text/plain, prompts user to “save as…”&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>AppleWatchアプリ開発におけるWCSessionのメソッド使用方法</title>
      <link>https://sankame.github.io/blog/2016-11-21-wcsession_sendmessage_slow_in_swift3/</link>
      <pubDate>Mon, 21 Nov 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-11-21-wcsession_sendmessage_slow_in_swift3/</guid>
      <description>

&lt;h3 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h3&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.6
[IDE] Xcode 8.1
[Swift] 3.0
&lt;/code&gt;&lt;/pre&gt;

&lt;h3 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;WCSession#sendMessage()を使ってAppleWatchからiPhoneアプリへメッセージを送っても、30秒ほどしないと通知されない。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;AppleWatch側&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-swift&#34;&gt;@IBAction func tapButton() {

    let message = [ &amp;quot;fromChild&amp;quot;: &amp;quot;AppleWatchからのメッセージ&amp;quot; ]

    WCSession.default().sendMessage(
            message
            , replyHandler: { reply in }
            , errorHandler: { error in }
    )

}
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;iPhone側&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-swift&#34;&gt;func session(_ session: WCSession, didReceiveMessage message: [String: Any], replyHandler: @escaping ([String: Any]) -&amp;gt; Void) {

    if let watchMessage = message[&amp;quot;fromChild&amp;quot;] as? String {
        print(watchMessage)
    }
    else{
        print(&amp;quot;error&amp;quot;)
    }

}
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;2-対策&#34;&gt;2. 対策&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;replyHandlerを設定するなど、いろいろ情報はあったのですが、結局 sendMessage() を諦め、updateApplicationContext() を使うことにしました。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;ただ updateApplicationContext() は状態が変わった際に通知するためのメソッドなので、二回目以降の値が変わらないとiPhone側のメソッドが呼び出されません。&lt;/p&gt;

&lt;p&gt;API仕様書には明記されていませんがそのような動きをしています。(「3. 参考ページ」参照)&lt;/p&gt;

&lt;p&gt;ですので、sendMessage() と同じように使いたい場合は、毎回変わる値をダミーとして設定すると実現できます。(多少強引ですが、下記例では時間を設定しています)&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;AppleWatch側&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-swift&#34;&gt;@IBAction func tapButton() {

    let dateFormatter = DateFormatter()
    dateFormatter.locale = Locale(identifier: &amp;quot;ja_JP&amp;quot;)
    dateFormatter.timeStyle = .medium
    dateFormatter.dateStyle = .medium

    let item: Dictionary&amp;lt;String, String&amp;gt; = [
        &amp;quot;message&amp;quot;: &amp;quot;AppleWatchからのメッセージ&amp;quot;
        , &amp;quot;date&amp;quot;: dateFormatter.string(from: NSDate() as Date)]

    let message = [ &amp;quot;fromChild&amp;quot;: item ]

    do{
        try WCSession.default().updateApplicationContext(message)
    }catch{
        print(error)
    }

}
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;iPhone側&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-swift&#34;&gt;func session(_ session: WCSession, didReceiveApplicationContext applicationContext: [String : Any]) {

    DispatchQueue.main.async { () -&amp;gt; Void in
        if let watchMessage = applicationContext[&amp;quot;fromChild&amp;quot;] as? Dictionary&amp;lt;String, String&amp;gt; {
            self.label.text = watchMessage[&amp;quot;message&amp;quot;]! as String
        }
    }
    return

}
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;3-参考ページ&#34;&gt;3. 参考ページ&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;https://developer.apple.com/reference/watchconnectivity/wcsession&#34; target=&#34;_blank&#34;&gt;API Reference WCSession&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>Type &#39;ViewController&#39; does not conform to protocol &#39;WCSessionDelegate&#39;</title>
      <link>https://sankame.github.io/blog/2016-11-15-wcsessiondelegate_in_swift3/</link>
      <pubDate>Tue, 15 Nov 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-11-15-wcsessiondelegate_in_swift3/</guid>
      <description>

&lt;h3 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h3&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.6
[IDE] Xcode 8.1
[Swift] 3.0
&lt;/code&gt;&lt;/pre&gt;

&lt;h3 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;iPhone &amp;amp; AppleWatchアプリを作成中にSwiftを2.2から3へバージョンアップしたところ、iPhone側で下記のエラーが出るようになりました。&lt;/p&gt;

&lt;p&gt;&lt;code&gt;Type &#39;ViewController&#39; does not conform to protocol &#39;WCSessionDelegate&#39;&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;2-解決&#34;&gt;2. 解決&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;ViewControllerに下記メソッドを追加するとエラーが消えました。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-swift&#34;&gt;public func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
}

func sessionDidBecomeInactive(_ session: WCSession) {
}

func sessionDidDeactivate(_ session: WCSession) {
    session.activate()
}
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;下記の仕様書にもこれらのメソッドは必須と書かれています。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://developer.apple.com/reference/watchconnectivity/wcsessiondelegate&#34; target=&#34;_blank&#34;&gt;[API Reference] WCSessionDelegate&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;3-参考ページ&#34;&gt;3. 参考ページ&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;https://forums.developer.apple.com/thread/63452&#34; target=&#34;_blank&#34;&gt;Migrating app from Swift 2.2 to 3.0 with WCSessionDelegate&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>型 UrlValidator は使用すべきではありません</title>
      <link>https://sankame.github.io/blog/2016-10-06-urlvalidator-is-deprecated/</link>
      <pubDate>Thu, 06 Oct 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-10-06-urlvalidator-is-deprecated/</guid>
      <description>

&lt;h3 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h3&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.4
[Java] 1.8.0
&lt;/code&gt;&lt;/pre&gt;

&lt;h3 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;久々にJavaで実装する機会がありました。標準ライブラリのjava.net.URLクラスは、URLとして解釈できない文字列を渡すと例外が投げられるので扱いずらいですね。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-java&#34;&gt;// IllegalArgumentException が発生。
url = URL.create(&amp;quot;不正なURL&amp;quot;);

&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;うーん、最初、RFCを見て「URLとして許可されている文字列だけ渡す」という処理を入れようか迷いました。&lt;/p&gt;

&lt;p&gt;が、一から組むのも骨が折れるなあ～と思って、困った時の&lt;a href=&#34;https://commons.apache.org/&#34; target=&#34;_blank&#34;&gt;Apache Commons&lt;/a&gt;を探してみたら、しめしめ&lt;a href=&#34;https://commons.apache.org/proper/commons-validator/download_validator.cgi&#34; target=&#34;_blank&#34;&gt;Apache Commons Validator&lt;/a&gt;にUrlValidatorクラスというのがありました。&lt;/p&gt;

&lt;p&gt;早速このクラスをimportして使おうと思ったら、下記の警告が出てしまいました。&lt;/p&gt;

&lt;p&gt;&lt;code&gt;型 UrlValidator は使用すべきではありません&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;2-解決&#34;&gt;2. 解決&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;調べてみたところ、①のパッケージは非推奨で②に変えたら警告も消えました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;&lt;font color=&#34;red&#34;&gt;①こっちは非推奨&lt;/font&gt;&lt;/p&gt;

&lt;p&gt;import org.apache.commons.validator.UrlValidator;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;font color=&#34;green&#34;&gt;②これを使って！&lt;/font&gt;&lt;/p&gt;

&lt;p&gt;import org.apache.commons.validator.routines.UrlValidator;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;3-参考&#34;&gt;3. 参考&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;http://stackoverflow.com/questions/14274919/the-type-urlvalidator-is-deprecated&#34; target=&#34;_blank&#34;&gt;The type UrlValidator is deprecated&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>composer updateでUndefined indexエラー</title>
      <link>https://sankame.github.io/blog/</link>
      <pubDate>Sat, 25 Jun 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/</guid>
      <description>

&lt;h3 id=&#34;0-環境&#34;&gt;0. 環境&lt;/h3&gt;

&lt;pre&gt;&lt;code&gt;[Laravel] 5.1
&lt;/code&gt;&lt;/pre&gt;

&lt;h3 id=&#34;1-問題&#34;&gt;1. 問題&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;composer update を実行すると下記エラーが発生しました。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;Writing lock file
Generating autoload files
&amp;gt; php artisan clear-compiled
  [ErrorException]
  Undefined index: HTTP_HOST
Script php artisan clear-compiled handling the post-update-cmd event returned with error code 1
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;2-解決策&#34;&gt;2. 解決策&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;HTTP_HOST でソースをgrepし、 /app/Http/routes.php 内の下記箇所を修正しました。&lt;/p&gt;

&lt;p&gt;(修正前)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-php&#34;&gt;$_SERVER[&#39;HTTP_HOST&#39;]
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;(修正後)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-php&#34;&gt;Request::server(&#39;HTTP_HOST&#39;)
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h3 id=&#34;3-参考&#34;&gt;3. 参考&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;http://stackoverflow.com/questions/34055053/composer-returning-mysterious-error&#34; target=&#34;_blank&#34;&gt;Composer returning mysterious error&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>watchOSのバージョンエラー</title>
      <link>https://sankame.github.io/blog/2016-05-09-watchos_devices_not_available/</link>
      <pubDate>Mon, 09 May 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-05-09-watchos_devices_not_available/</guid>
      <description>

&lt;h1 id=&#34;0-環境&#34;&gt;0.環境&lt;/h1&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.4
[IDE] Xcode 7.3
[iOS] 9.3
[watchOS] 2.1

※MacとiPhoneはUSBケーブルで接続。
※iPhoneとAppleWatchはペアリング済み。
&lt;/code&gt;&lt;/pre&gt;

&lt;h1 id=&#34;1-問題-文中-は読み替えて下さい&#34;&gt;1.問題 (文中【】は読み替えて下さい)&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;AppleWatch用のサンプルプログラムを実機で動かそうとしたら、下記3つのエラーが出ました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;&lt;b&gt;エラー1&lt;/b&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;A build only device cannot be used to run this target.&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;No supported watchOS devices are available. Pair a device to run your application or choose a simulated device as the destination.&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;b&gt;エラー2&lt;/b&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;Could not launch &#39;【アプリ名】 WatchKit App&#39; on 【iPhoneデバイス名】&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;OS version lower than deployment target&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;b&gt;エラー3&lt;/b&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;Can&#39;t install application&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;【アプリ名】 WatchKit App.app cannot be installed on 【AppleWatchデバイス名】.&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;The Info.plist in 【アプリ名】 WatchKit App.app specifies a minimum iOS version of 2.2 which is too high to be installed on 【AppleWatchデバイス名】.&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-手順&#34;&gt;2.手順&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;AppleWatch実機のOSバージョンとXcodeで指定しているターゲットバージョンが異なっていたことが原因でした。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;下記の変更を行ったところ、問題なく実機にインストールされました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;プロジェクトナビゲーターで該当プロジェクトを選択。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&amp;ldquo;Build Settings&amp;rdquo;を選択してターゲットを表示。&lt;font color=&#34;orange&#34;&gt;(A-1、A-2)&lt;/font&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;watchOSのバージョンを実機に合わせます。(僕の場合2.2→2.1に変更) &lt;font color=&#34;orange&#34;&gt;(A-3)&lt;/font&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;(&amp;ldquo;WatchKit App&amp;rdquo;を変更すると、&amp;rdquo;WatchKit Extension&amp;rdquo;も連動して変わります)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/watchos_devices_not_available/change_the_version.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;100%&#34; height=&#34;100%&#34; title=&#34;ターゲットバージョン変更&#34; alt=&#34;ターゲットバージョン変更&#34; src=&#34;https://sankame.github.io/assets/posts/img/watchos_devices_not_available/change_the_version.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;いくつかサンプルプログラムを作っていた中で、なぜこれだけターゲットバージョンがデフォルトで2.2になっていたのか不明です。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>AppleWatch 開発中アプリのインストール制限を回避</title>
      <link>https://sankame.github.io/blog/2016-05-08-unable_to_add_app_id/</link>
      <pubDate>Sun, 08 May 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-05-08-unable_to_add_app_id/</guid>
      <description>

&lt;p&gt;&lt;span&gt;&amp;nbsp;&lt;/span&gt;&lt;span style=&#34;color:#d14; float:right;&#34;&gt;更新日：2016年05月11日&lt;/span&gt;&lt;/p&gt;

&lt;h2 id=&#34;0-環境&#34;&gt;0.環境&lt;/h2&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.4
[IDE] Xcode 7.3
[iOS] 9.3
[watchOS] 2.1

※ MacとiPhoneはUSBケーブルで接続。
※ iPhoneとAppleWatchはペアリング済み。
&lt;/code&gt;&lt;/pre&gt;

&lt;h2 id=&#34;1-問題&#34;&gt;1.問題&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;AppleWatch用のサンプルプログラムを実機で動かしていたら、下記エラーが出るようになりました。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/error_message.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;50%&#34; height=&#34;50%&#34; title=&#34;Unable to add App ID&#34; alt=&#34;Unable to add App ID&#34; src=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/error_message.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;Unable to add App ID because the &#39;10&#39; App ID limit in &#39;7&#39; days has been exceeded.&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;「7日間で10個のアプリID制限を超えたため、アプリIDを追加できません」か・・・&lt;/p&gt;

&lt;p&gt;Webで調べたところ回避策は2つありましたが、簡単そうな①をやってみました。&lt;/p&gt;

&lt;p&gt;&lt;code&gt;① Xcodeに新しいApple IDを追加して、それを使ってインストール&lt;/code&gt;&lt;/p&gt;

&lt;p&gt;&lt;code&gt;② 既に実機にインストール済みのアプリの Bundle ID を使い回す&lt;/code&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h2 id=&#34;2-手順&#34;&gt;2.手順&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;(画像内の【】は読み替えて下さい)&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;Xcodeに新しい Apple ID を追加&lt;/p&gt;

&lt;p&gt;アプリの[General]-[Team]をクリックし &amp;ldquo;Add an Account&amp;rdquo; を選択します。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/add_account.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;70%&#34; height=&#34;70%&#34; title=&#34;AppleID追加&#34; alt=&#34;AppleID追加&#34; src=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/add_account.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;Xcodeに追加するApple IDを要求されますが、未作成なので&amp;rdquo;Create Apple ID&amp;rdquo;をクリック。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/required_to_sign_in.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;50%&#34; height=&#34;50%&#34; title=&#34;サインイン&#34; alt=&#34;サインイン&#34; src=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/required_to_sign_in.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;必要項目を入力していくと、最後に「入力したメールアドレス宛てにメールを送ったから確認してね」と表示されます。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/verification_for_the_email.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;50%&#34; height=&#34;50%&#34; title=&#34;email確認&#34; alt=&#34;email確認&#34; src=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/verification_for_the_email.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;上記メールアドレス宛てに確認メールが届いているので、メール本文内の&amp;rdquo;今すぐ確認&amp;rdquo;をクリックして登録を完了させます。&lt;/p&gt;

&lt;p&gt;再度、手順の最初に戻って、&amp;rdquo;Add an Account&amp;rdquo;を選択し、登録したての新しいApple IDを入力します。&lt;/p&gt;

&lt;p&gt;新しいApple IDを選択すると下記のメッセージが出るので&amp;rdquo;Fix Issue&amp;rdquo;→&amp;rdquo;Choose&amp;rdquo;をクリックして解消します。&lt;/p&gt;

&lt;p&gt;&lt;font color=&#34;red&#34;&gt;【注意】ダイアログ内のプルダウンで新しいApple IDが選択されているか再度確認して下さい&lt;/font&gt;&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/fix_the_issue.jpg&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;50%&#34; height=&#34;50%&#34; title=&#34;Fix Issue&#34; alt=&#34;Fix Issue&#34; src=&#34;https://sankame.github.io/assets/posts/img/unable_to_add_app_id/fix_the_issue.jpg&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;そのままアプリをビルド＆起動すると署名関係の警告が何度か出ますが、&amp;rdquo;Fix issue&amp;rdquo; をクリックしていくと実機でアプリを実行できました。&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h2 id=&#34;参考ページ&#34;&gt;参考ページ&lt;/h2&gt;

&lt;ul&gt;
&lt;li&gt;①&lt;a href=&#34;http://stackoverflow.com/questions/36922738/how-to-get-around-apple-app-id-insanity&#34; target=&#34;_blank&#34;&gt;How to get around Apple app ID insanity&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>pod setup で SSLエラー</title>
      <link>https://sankame.github.io/blog/2016-04-21-pod_setup_ssl_error/</link>
      <pubDate>Thu, 21 Apr 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-04-21-pod_setup_ssl_error/</guid>
      <description>

&lt;h1 id=&#34;0-環境&#34;&gt;0.環境&lt;/h1&gt;

&lt;pre&gt;&lt;code&gt;[OS] Mac OS X 10.11.4
[cocoapods] 1.0.0.beta.6
&lt;/code&gt;&lt;/pre&gt;

&lt;h1 id=&#34;1-エラー内容&#34;&gt;1.エラー内容&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;pod setupを実行すると下記エラーが発生。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;$ pod setup --verbose

Setting up CocoaPods master repo

Cloning spec repo `master` from `https://github.com/CocoaPods/Specs.git` (branch `master`)
  $ /usr/bin/git clone https://github.com/CocoaPods/Specs.git master
  Cloning into &#39;master&#39;...
  fatal: unable to access &#39;https://github.com/CocoaPods/Specs.git/&#39;: SSLRead() return error -9806
[!] /usr/bin/git clone https://github.com/CocoaPods/Specs.git master

Cloning into &#39;master&#39;...
fatal: unable to access &#39;https://github.com/CocoaPods/Specs.git/&#39;: SSLRead() return error -9806

        ：
     【省略】
        ：
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-解決方法&#34;&gt;2.解決方法&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Safariで下記URLにアクセス。(他のブラウザではダメ)&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://github.com&#34; target=&#34;_blank&#34;&gt;&lt;a href=&#34;https://github.com&#34;&gt;https://github.com&lt;/a&gt;&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;再度、下記を実行するとwifi経由で1時間以上経ってからようやく成功…&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;$ pod setup --verbose

Setting up CocoaPods master repo

        ：
     【省略】
        ：

Setup completed

&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;参考ページ&#34;&gt;参考ページ&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href=&#34;http://stackoverflow.com/questions/20939105/pod-install-returns-fatal-error-ssl-certificate-issue&#34; target=&#34;_blank&#34;&gt;pod install returns fatal error: SSL certificate issue?&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
    <item>
      <title>Rundeckでssh接続エラー</title>
      <link>https://sankame.github.io/blog/2016-02-15-rundeck_ssh_error/</link>
      <pubDate>Mon, 15 Feb 2016 00:00:00 +0000</pubDate>
      
      <guid>https://sankame.github.io/blog/2016-02-15-rundeck_ssh_error/</guid>
      <description>

&lt;h1 id=&#34;0-環境&#34;&gt;0.環境&lt;/h1&gt;

&lt;p&gt;&lt;code&gt;[CentOS] 5.6&lt;/code&gt;&lt;/p&gt;

&lt;h1 id=&#34;1-問題&#34;&gt;1.問題&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;文中、画像内の【】は読み替えて下さい。&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;下記ページ内の「リモートサーバでの実行」にそってRundeckのリモート接続を試してみました。&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;http://dev.classmethod.jp/server-side/server/try-rundeck-job/&#34; target=&#34;_blank&#34;&gt;&amp;ldquo;ジョブスケジューラ「Rundeck」を試してみる&amp;rdquo;&lt;/a&gt;&lt;/p&gt;

&lt;p&gt;ところが、このエラーが発生…&lt;/p&gt;

&lt;p&gt;&lt;a href=&#34;https://sankame.github.io/assets/posts/img/rundeck_ssh_error/rundeck_ssh_error.png&#34; target=&#34;_blank&#34;&gt;&lt;img width=&#34;100%&#34; height=&#34;100%&#34; title=&#34;RundeckSSHエラー&#34; alt=&#34;RundeckSSHエラー&#34; src=&#34;https://sankame.github.io/assets/posts/img/rundeck_ssh_error/rundeck_ssh_error.png&#34;&gt;&lt;/a&gt;&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;上記画像内のテキストはこちら。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-text&#34;&gt;22:34:10 remote 1. hogeと表示。 Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.
22:34:10 Failed: AuthenticationFailure: Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.
22:34:10 localhost Execution failed: 21: [Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [remote: AuthenticationFailure: Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.]}, Node failures: {remote=[AuthenticationFailure: Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.]}, status: failed]
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;2-デバッグ&#34;&gt;2.デバッグ&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;まずはRundeckで対象プロジェクトの[Prepare and Run&amp;hellip;]-[Log level] を&amp;rdquo;Debug&amp;rdquo;に変更し、エラーの詳細を表示させます。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;詳細を見ると、最初の公開鍵認証で失敗し、さらにその後のパスワード認証でも失敗しているようです。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-text&#34;&gt;22:37:48 Authentications that can continue: publickey,password,keyboard-interactive
22:37:48 Next authentication method: publickey
22:37:48 Authentications that can continue: password,keyboard-interactive
22:37:48 Next authentication method: password
22:37:48 Disconnecting from 【IPアドレス】 port 22
22:37:48 Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.
22:37:48 Failed: AuthenticationFailure: Authentication failure connecting to node: &amp;quot;remote&amp;quot;. Make sure your resource definitions and credentials are up to date.
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;p&gt;おかしいなあ、公開鍵認証で接続できる設定なんだけど…&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;Rundeck以前にそもそもsshで失敗しているようなので、ssh単体で実行してみます。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;ssh -i $(awk &#39;/framework.ssh.keypath = / {print $3}&#39; /etc/rundeck/framework.properties) rundeck@【IPアドレス】
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;さらにコマンドを簡素化して実行してみると、パスワード未設定の公開鍵を使っているのになぜかパスワードを聞かれます。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;[22:16:34][rundeck@【サーバー名】 ~/.ssh]$ ssh -i /var/lib/rundeck/.ssh/id_rsa rundeck@【IPアドレス】
rundeck@【IPアドレス】&#39;s password: 
Permission denied, please try again.
rundeck@【IPアドレス】&#39;s password: 
Permission denied, please try again.
rundeck@【IPアドレス】&#39;s password: 
Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;3-原因と解決&#34;&gt;3.原因と解決&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;色々試した結果、接続先サーバーにある公開鍵ファイルのパーミッションが原因でした。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;&amp;lt;接続元&amp;gt;&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-sh&#34;&gt;ssh -i /var/lib/rundeck/.ssh/id_rsa rundeck@【IPアドレス】
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&amp;lt;接続先&amp;gt;&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;パーミッションが緩い分には問題ないだろうと思っていたら、&lt;font color=&#34;red&#34;&gt;厳密に600じゃないとダメ&lt;/font&gt;なんですね…&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;code&gt;chmod 600 authorized_keys&lt;/code&gt; を実行し、下記のように修正したらssh接続もRundeckからのリモート接続も成功しました。&lt;/p&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;(変更前)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;-rw-rw-r--. 1 rundeck rundeck 403 2月 4 15:58 2016 authorized_keys
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;(変更後)&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-console&#34;&gt;-rw-------. 1 rundeck rundeck 403 2月 4 15:58 2016 authorized_keys
&lt;/code&gt;&lt;/pre&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;h1 id=&#34;参考になった記事&#34;&gt;参考になった記事&lt;/h1&gt;

&lt;ul&gt;
&lt;li&gt;&lt;p&gt;&lt;a href=&#34;http://qiita.com/gotohiro55/items/36a22516de2b381b3c6e&#34; target=&#34;_blank&#34;&gt;SSHの鍵認証設定&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;

&lt;li&gt;&lt;p&gt;&lt;a href=&#34;https://rundeck.lighthouseapp.com/projects/59277/tickets/345-add-resource&#34; target=&#34;_blank&#34;&gt;Rundeck Development&lt;/a&gt;&lt;/p&gt;&lt;/li&gt;
&lt;/ul&gt;
</description>
    </item>
    
  </channel>
</rss>